### DISCUSSION TITLE
[readable/writable]StreamFrom[Reader/Writer] vs FsFile.[readable/writable]

### DISCUSSION THREAD

**Discussion by davbrito** [denoland/deno#17102] [Q&A] [answered]
There is currently two ways of getting a Web Stream from a `FsFile` object. Since a `FsFile` is a Reader and a Writer you could use the std helpers to do the task:

```ts
import { readableStreamFromReader } from "https://deno.land/std@0.168.0/std/streams/readable_stream_from_reader.ts";
import { writableStreamFromWriter } from "https://deno.land/std@0.168.0/std/streams/writable_stream_from_writer.ts";

const source = await Deno.open("source.txt", { read: true });
const dest = await Deno.open("dest.txt", { write: true });

await readableStreamFromReader(source).pipeTo(writableStreamFromWriter(dest));
// OR, in a web server context
new Response(readableStreamFromReader(source))
request.body.pipeTo(writableStreamFromWriter(dest))
```

The other way is simpler. Use just the `.readable` or `.writable` properties from the FsFile object:
```ts
const source = await Deno.open("source.txt", { read: true });
const dest = await Deno.open("dest.txt", { write: true });

await source.readable.pipeTo(dest.writable);
// OR, in a web server context
new Response(source.readable)
request.body.pipeTo(dest.writable)
```

Which one of those would be the recommended way to do that. I have seen the first pattern used all the way from the docs. But the latter not. Is there a reason for that? Which one would be more reliable?

**Comment by ayame113** [comment, marked as answer]
In Deno, Go-style streaming (`Deno.reader/writer`) used to be the norm. However, the Web Stream API was later implemented instead of Go-style streaming, and it is now recommended to use it.

ref: https://github.com/denoland/deno/issues/9795#issuecomment-874347722, https://github.com/denoland/deno_std/issues/1986

`readableStreamFromReader` and `writableStreamFromWriter` are utilities provided (to my understanding) to migrate from Go-style streaming to Web Stream API.
Therefore, I think it is recommended to write in the second style (using only Web Stream API) unless there is a special reason.

**Reply by crowlKats** [reply]
yes, using the getter is the recommended way. the former way is just legacy by now.

additionally, to answer the question:
> Is there a reason for that?

the main one being is the former way was around for quite a while longer, unlike the streams properties which were added quite later

**Reply by davbrito** [reply]
Thanks for the quick response :tada:
